src/pin.rs

```diff
@@ -0,0 +1,138 @@
+
+use crate::common::Generic;
+use crate::error::MockError;
+
+use embedded_hal::digital::v2::{OutputPin, InputPin};
+
+/// MockPin transaction
+#[derive(Debug)]
+pub struct Transaction {
+    kind: TransactionKind,
+    err: Option<MockError>,
+}
+
+impl Transaction {
+    /// Create a new pin transaction
+    pub fn new(kind: TransactionKind, err: Option<MockError>) -> Transaction {
+        Transaction{kind, err}
+    }
+}
+
+/// MockPin transaction kind, either Get or Set a bool
+#[derive(Debug, PartialEq)]
+pub enum TransactionKind {
+    /// Set(true) for set_high or Set(false) for set_low
+    Set(bool),
+    /// Get(true) for high value or Get(false) for low value
+    Get(bool),
+}
+
+/// Mock Pin implementation
+pub type Mock<'a> = Generic<'a, Transaction>;
+
+/// Single digital push-pull output pin
+impl <'a>OutputPin for Mock<'a> {
+    /// Error type
+    type Error = MockError;
+
+    /// Drives the pin low
+    fn set_low(&mut self) -> Result<(), Self::Error> {
+        let Transaction{kind, err} = self.next().expect("no expectation for pin::set_low call");
+
+        assert_eq!(*kind, TransactionKind::Set(false), "expected pin::set_low");
+        
+        match err {
+            Some(e) => Err(e.clone()),
+            None => Ok(()),
+        }
+    }
+
+    /// Drives the pin high
+    fn set_high(&mut self) -> Result<(), Self::Error> {
+        let Transaction{kind, err} = self.next().expect("no expectation for pin::set_high call");
+
+        assert_eq!(*kind, TransactionKind::Set(true), "expected pin::set_high");
+        
+        match err {
+            Some(e) => Err(e.clone()),
+            None => Ok(()),
+        }
+    }
+}
+
+impl <'a>InputPin for Mock<'a> {
+    /// Error type
+    type Error = MockError;
+
+    /// Is the input pin high?
+    fn is_high(&self) -> Result<bool, Self::Error> {
+        let mut s = self.clone();
+
+        let Transaction{kind, err} = s.next().expect("no expectation for pin::is_high call");
+
+        if let Some(e) = err { 
+            Err(e.clone())
+        } else if let TransactionKind::Get(v) = kind {
+            Ok(*v == true)
+        } else {
+            panic!("expected pin::is_high (actual: {:?}", kind);
```

Comment:
Why the check-and-panic here and an assertion for the `set_` methods? Maybe the same approach could be used for both? I think the error messages differ otherwise, which might confuse users.

Reply:
For the set case we can check for `Set(true)` or `Set(false)` for `set_high` and `set_low`, which makes the assert easy, for the get case `is_high` and `is_low` are both valid so I ended up with that. Open to suggestions if you have a better idea?

Reply:
Added an `is_get` method and swapped to `assert` with `unreachable`... does that seem reasonable?